kratos/python_scripts/sympy_fe_utilities.py

```diff
@@ -1,466 +1,522 @@
+    - testfunc -- The test functions
+    - dofs -- The dofs vectors
+    - do_simplifications -- If apply simplifications
     """
     rhs = Compute_RHS(functional, testfunc, do_simplifications)
     lhs = Compute_LHS(rhs, testfunc, dofs, do_simplifications)
     return rhs,lhs
 
-def OutputVector(r, name, mode="python", initial_tabs=3, max_index=30, replace_indices=True, assignment_op="="):
-    """ This method converts into text the RHS vector
+# Output functions
+def _Indentation(indentation_level):
+    """Returns the indentation string."""
+    return "    " * indentation_level
+
+def _CodeGen(language, value):
+    return  {
+        "c"     : sympy.ccode,
+        "python": sympy.pycode
+    }[language](value)
+
+def _VariableDeclaration(language, variable_name, variable_expression):
+    """"
+    Returns the variable declaration, without indentation nor suffix.
+
+    The expression must have been turned into code already.
+    """
+    return  {
+        "c"     : "const double {name} = {expr}",
+        "python": "{name} = {expr}"
+    }[language].format(name=variable_name, expr=variable_expression)
+
+def _Suffix(language):
+    """Returns the endline suffix."""
+    return  {
+        "c"     : ";\n",
+        "python": "\n"
+    }[language]
+
+def _ReplaceIndices(language, expression):
+    """Replaces array access with underscored variable.
+
+    For matrices: `variable[3,7]` becomes `variable_3_7`
+    For vectors:  `variable[3]` becomes `variable_3`
+
+    The accessor for matrices is chosen according to the language (`[]` vs. `()`)
+    """
+    #Matrices
+    pattern = r"\[(\d+),(\d+)\]" if language == 'python' else r"\((\d+),(\d+)\)"
+    replacement = r"_\1_\2"
+    expression = re.sub(pattern, replacement, expression)
+
+    # Vectors
+    pattern = r"\[(\d+)\]"
+    replacement = r"_\1"
+    expression = re.sub(pattern, replacement, expression)
+
+    return expression
+
+def OutputScalar(scalar_expression, name, language, indentation_level=0, replace_indices=True, assignment_op="="):
+    """
+    This function generates code to assign to a (pre-declared) scalar
+
+    Keyword arguments:
+    - scalar_expression -- A scalar
+    - name -- The name of the variables
+    - language -- The language of output
+    - indentation_level -- The number of tabulations considered
+    - replace_indices -- Set to `True` to replace matrix[i,j] with matrix_i_j (And similarly for vectors)
+    - assignment_op -- The assignment operation
+    """
+    prefix = _Indentation(indentation_level)
+    suffix = _Suffix(language)
+
+    fmt = prefix + "{var}{op}{expr}" + suffix
+
+    expression = _CodeGen(language, scalar_expression)
+    outstring = fmt.format(var=name, op=assignment_op, expr=expression)
+
+    if replace_indices:
+        outstring = _ReplaceIndices(language, outstring)
+
+    return outstring
+
+def OutputVector(vector_expression, name, language="python", indentation_level=0, replace_indices=True, assignment_op="="):
```

Comment:
```suggestion
def OutputVector(vector_expression, name, language, indentation_level=0, replace_indices=True, assignment_op="="):
```
I'd always ask for this value.

Reply:
Indeed, this is what we do in the scalar and matrix methods.

Reply:
Agreed.